nupic/research/frameworks/pytorch/model_utils.py

```diff
@@ -178,9 +178,13 @@ def evaluate_model(
     :rtype: dict
     """
     model.eval()
-    loss = 0
-    correct = 0
     total = 0
+
+    # Perform accumulation on device, avoid paying performance cost of
+    # tensor.to(device) and .item().
+    loss = None
+    correct = None
```

Comment:
It would be simpler to create the tensors here instead of checking for `None` later. Something like this:
```
loss = torch.tensor(0.)
correct = torch.tensor(0.)
```
This way we could leave most of the code alone removing the `.item()` from the loop and only call `.item()` at the end.
```
return {
        "total_correct": correct.item(),
        "mean_loss": loss.item() / total if total > 0 else 0,
```

Reply:
That's a good point, although we will need to move the tensors to the device. So the lines will need to be:

```
loss = torch.tensor(0.).to(device)
correct = torch.tensor(0.).to(device)
```

I actually tried this last week and found that the calls to `tensor.to` had some real overhead. But I'm going to try it again; I may have overlooked that there's another call to `.to` in this function:

```
# Maybe the actual culprit
data = data.to(device, non_blocking=async_gpu)
target = target.to(device, non_blocking=async_gpu)
```

Reply:
You can also try passing the `device` in the constructor.
```
loss = torch.tensor(0., device=device)
correct = torch.tensor(0., device=device)
```